tests/numeric.rs

```diff
@@ -0,0 +1,139 @@
+use hvmc::{
+  ast::{show_net, Book},
+  run,
+  run::NumericOp,
+};
+use insta::{assert_debug_snapshot, assert_snapshot};
+use loaders::*;
+
+mod loaders;
+
+fn op_net(lnum: u32, op: NumericOp, rnum: u32) -> Book {
+  parse_core(&format!("@main = root & <#{lnum} <#{rnum} root>> ~ #{op}"))
+}
+
+#[test]
+fn test_add() {
+  let net = op_net(10, run::ADD, 2);
+  let (rnet, net) = normal(net, 16);
+  assert_snapshot!(show_net(&net), @"#12");
+  assert_debug_snapshot!(rnet.rewrites(), @"5");
+}
+
+#[test]
+fn test_sub() {
+  let net = op_net(10, run::SUB, 2);
+  let (_, net) = normal(net, 16);
+  assert_snapshot!(show_net(&net), @"#8");
+}
+
+#[test]
+fn test_mul() {
+  let net = op_net(10, run::MUL, 2);
+  let (_, net) = normal(net, 16);
+  assert_snapshot!(show_net(&net), @"#20");
+}
+
+#[test]
+fn test_div() {
+  let net = op_net(10, run::DIV, 2);
+  let (_, net) = normal(net, 16);
+  assert_snapshot!(show_net(&net), @"#5");
+}
+
+#[test]
+fn test_mod() {
+  let net = op_net(10, run::MOD, 2);
+  let (_, net) = normal(net, 16);
+  assert_snapshot!(show_net(&net), @"#0");
+}
+
+#[test]
+fn test_eq() {
+  let net = op_net(10, run::EQ, 2);
+  let (_, net) = normal(net, 16);
+  assert_snapshot!(show_net(&net), @"#0");
+}
+
+#[test]
+fn test_ne() {
+  let net = op_net(10, run::NE, 2);
+  let (_, net) = normal(net, 16);
+  assert_snapshot!(show_net(&net), @"#1");
+}
+
+#[test]
+fn test_lt() {
+  let net = op_net(10, run::LT, 2);
+  let (_, net) = normal(net, 16);
+  assert_snapshot!(show_net(&net), @"#0");
+}
+
+#[test]
+fn test_gt() {
+  let net = op_net(10, run::GT, 2);
+  let (_, net) = normal(net, 16);
+  assert_snapshot!(show_net(&net), @"#1");
+}
+
+#[test]
+fn test_and() {
+  let net = op_net(10, run::AND, 2);
+  let (_, net) = normal(net, 16);
+  assert_snapshot!(show_net(&net), @"#2");
+}
+
+#[test]
+fn test_or() {
+  let net = op_net(10, run::OR, 2);
+  let (_, net) = normal(net, 16);
+  assert_snapshot!(show_net(&net), @"#10");
+}
+
+#[test]
+fn test_xor() {
+  let net = op_net(10, run::XOR, 2);
+  let (_, net) = normal(net, 16);
+  assert_snapshot!(show_net(&net), @"#8");
+}
+
+#[test]
+fn test_not() {
+  let net = op_net(0, run::NOT, 256);
+  let (rnet, net) = normal(net, 16);
+  assert_snapshot!(show_net(&net), @"#16776959");
+  assert_debug_snapshot!(rnet.rewrites(), @"4");
+}
+
+#[test]
+fn test_lsh() {
+  let net = op_net(10, run::LSH, 2);
+  let (_, net) = normal(net, 16);
+  assert_snapshot!(show_net(&net), @"#40");
+}
+
+#[test]
+fn test_rsh() {
+  let net = op_net(10, run::RSH, 2);
+  let (_, net) = normal(net, 16);
+  assert_snapshot!(show_net(&net), @"#2");
+}
+
+#[test]
+/// Division by zero always return the value of 0xFFFFFF,
+/// that is read as the unsigned integer `16777215`
+fn test_div_by_0() {
+  let net = op_net(9, run::DIV, 0);
+  let (rnet, net) = normal(net, 16);
+  assert_snapshot!(show_net(&net), @"#16777215");
+  assert_debug_snapshot!(rnet.rewrites(), @"5");
+}
+
+#[test]
+fn test_chained_ops() {
```

Comment:
is there a way to test if it is actually doing chained ops? instead of doing one op per interaction? if you don't know or if there's a way but it takes a lot of time, then you don't need to do this(maybe add a comment only)

Reply:
From my understanding, there is no way. As Taelin changed the function `op2n` to perform this optimization, but the number of rewrites is maintained the same. I believe the number of calls to the `interact` function is reduced, but we don`t have a way to measure that.
![image](https://github.com/HigherOrderCO/hvm-core/assets/19491161/e2e6f141-ed33-4025-a2a2-1f63795be72d)

Reply:
I tested this file with both versions of the op2n function, to make sure, and the new version results in 3 less calls.

Reply:
If you call the `interact` function directly, you can check `net.oper` increases by more than one in a single call

Reply:
how did you measure it made 3 less calls? but yeah, it seems it would be hard to detect that with tests, I think you can just comment it though: `// TODO: we lack a way to check if it's actually doing the chained ops optimization, or if it's doing one op per interaction`.

actually, maybe we could add another stat field for that, but I wouldn't do it right now, the comment is good enough